### PR TITLE
Oud support

### PR DESCRIPTION
**Pull request by shero0x1337**
- Support Multi-asset borrowing with different parameters

### CODE REVIEW

File: src/TempleLineOfCredit.sol [hunk 1]
```diff
@@ -98,125 +127,177 @@ contract TempleLineOfCredit is Ownable, Operators {
...
 
+    function getDebtTokenPrice(TokenPrice _price) public view returns (uint256 price, uint256 precision) {
+
+        if (_price == TokenPrice.STABLE) {
+            return (10000, 10000);
+        } else if (_price == TokenPrice.TPI) {
+            // Get Token Price from redemeer
+            uint256 tpiPrice = IOudRedeemer(oudRedeemer).treasuryPriceIndex();
+            return (tpiPrice, 10000);
+        }
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Should error if neither case?

File: src/TempleLineOfCredit.sol [hunk 2]
```diff
@@ -98,125 +127,177 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+    }
+
     /**
      * @dev Allows operator to depoist debt tokens
      * @param amount is the amount to deposit
      */
-    function depositDebt(address account, uint256 amount) external onlyOperators{
+    function depositDebt(address debtToken, address account, uint256 amount) external onlyOperators{
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Kinda weird that this is called `depositDebt()`. Maybe 
depositReserves()
removeReserves()
?

File: src/TempleLineOfCredit.sol [hunk 3]
```diff
@@ -98,125 +127,177 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+    function borrow(address[] memory tokens, uint256[] memory borrowAmounts) external {
 
-        uint256 debtAmount =  positions[msg.sender].debtAmount;
-        if (borrowAmount > debtBalance) {
-            revert InsufficentDebtToken(debtBalance, borrowAmount);
-        }
-        if (debtAmount != 0) {
-            debtAmount = getTotalDebtAmount(msg.sender);
-        }
+        for (uint256 i =0; i < tokens.length; i++) {
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: A few minor gas savings with:

1.
```
for (uint256 i; i < tokens.length; ++i) {
```
Same for other loops below too.

2. Move the declaration of vars above the loop, ie:
```
address debtToken;
uint256 borrowAmount;
DebtToken memory debtTokenInfo;
...
for (uint256 i; i < tokens.length; ++i) {
    ...
}
```
So they don't have to be re-allocated each iteration.


Should also check that `tokens.length == borrowAmounts.length`

File: src/TempleLineOfCredit.sol [hunk 4]
```diff
@@ -98,125 +127,177 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+            }
+
+            uint256 borrowCapacity = _maxBorrowCapacity(positions[msg.sender].collateralAmount, debtTokenInfo.tokenPrice, debtTokenInfo.minCollateralizationRatio) - debtAmount;
+            debtAmount += borrowAmount;
+
+            if (debtAmount > borrowCapacity) {
+                revert InsufficentCollateral(debtToken, borrowCapacity, debtAmount);
+            }
+
+            TokenPosition storage userPosition = positions[msg.sender].tokenPosition[debtToken];
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Can use this storage for the `debtAmount` above too rather than pulling it out twice

File: src/TempleLineOfCredit.sol [hunk 5]
```diff
@@ -98,125 +127,177 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+
+            userPosition.debtAmount = debtAmount;
+            userPosition.createdAt = block.timestamp;
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: The timestamp changes over time, so it's not really `createdAt`. Perhaps call `lastUpdatedAt`?

File: src/TempleLineOfCredit.sol [hunk 6]
```diff
@@ -98,125 +127,177 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+    function maxBorrowCapacity(address debtToken, address account) public view returns(uint256) {
+        DebtToken memory debtTokenInfo = debtTokens[debtToken];
+        return  _maxBorrowCapacity(positions[account].collateralAmount, debtTokenInfo.tokenPrice, debtTokenInfo.minCollateralizationRatio);
     }
 
 
-    function _maxBorrowCapacity(uint256 collateralAmount) internal view returns (uint256) {
-        return collateralAmount * collateralPrice * 10000 / debtPrice / minCollateralizationRatio;
+    function _maxBorrowCapacity(uint256 collateralAmount, TokenPrice debtPrice,  uint256 minCollateralizationRatio) internal view returns (uint256) {
+        (uint256 debtTokenPrice, uint256 debtPercision) = getDebtTokenPrice(debtPrice);
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: typo: `debtPercision` should be `debtPrecision`

File: src/TempleLineOfCredit.sol (hunk 2 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol]: 👍

File: src/TempleLineOfCredit.sol (hunk 3 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol]: 👍

File: src/TempleLineOfCredit.sol [hunk 7]
```diff
@@ -8,85 +8,115 @@ import {SafeERC20} from "openzeppelin-contracts/token/ERC20/utils/SafeERC20.sol"
...
+
+interface IOudRedeemer {
+    function treasuryPriceIndex() external view returns (uint256);
+}
+
 contract TempleLineOfCredit is Ownable, Operators {
 
     using SafeERC20 for IERC20;
 
+    /// @notice debt position on all tokens
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Maybe the comment should be:
`/// A user's posted collateral, debt positions across all debt tokens.`

And maybe the `TokenPosition` struct and param should be renamed to `DebtPosition`?

File: src/TempleLineOfCredit.sol [hunk 8]
```diff
@@ -8,85 +8,115 @@ import {SafeERC20} from "openzeppelin-contracts/token/ERC20/utils/SafeERC20.sol"
...
-    /// @notice Supported collateral token address
-    IERC20 public immutable collateralToken;
+    enum TokenType {
+        MINT,
+        TRANSFER
+    }
 
-    /// @notice Collateral token price
-    uint256 public collateralPrice;
+    enum TokenPrice {
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:43]: Can you add a comment to say what STABLE vs TPI does?

File: src/TempleLineOfCredit.sol [hunk 9]
```diff
@@ -8,85 +8,115 @@ import {SafeERC20} from "openzeppelin-contracts/token/ERC20/utils/SafeERC20.sol"
...
+    /// @notice Address to send bad debt collateral
+    address public debtCollector;
+
     /// @notice Mapping of user positions
     mapping(address => Position) public positions;
 
-    event DepositDebt(uint256 amount);
-    event RemoveDebt(uint256 amount);
+    event SetCollateralPrice(TokenPrice price);
+    event SetDebtCollector(address debtCollector);
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:88]: A lot of these should be `indexed` event params: https://docs.soliditylang.org/en/v0.8.18/contracts.html#events

File: src/TempleLineOfCredit.sol [hunk 10]
```diff
@@ -8,85 +8,115 @@ import {SafeERC20} from "openzeppelin-contracts/token/ERC20/utils/SafeERC20.sol"
...
+    function mint(address to, uint256 amount) external;
+    function burn(address from, uint256 amount) external;
+
+}
+
+interface IOudRedeemer {
+    function treasuryPriceIndex() external view returns (uint256);
+}
+
 contract TempleLineOfCredit is Ownable, Operators {
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:21]: I think we agreed this should be a UUPS upgradeable contract, no? Especially given the amount of spec change recently, I think this is a necessity...

File: src/TempleLineOfCredit.sol [hunk 11]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+        debtTokens[token] = newDebtToken;
+        debtTokenList.push(token);
+        emit AddDebtToken(token);
     }
 
+
+    function removeDebtToken(address debtToken) external onlyOperators {
+        if (!debtTokens[debtToken].isAllowed) revert Unsupported(debtToken);
+        delete debtTokens[debtToken];
+        /// TODO: Remove from the debtTokenList
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: This is still missing. Should be fine to just:

```
for (uint256 i; i < debtTokenList.length; ++i) {
    if (debtTokenList[i] == debtToken) {
        // Switch the last item into this place then pop off the end.
        debtTokenList[i] = debtTokenList[debtTokenList.length - 1];
        debtTokenList.pop();
        break;
    }
}
```

I wouldn't think order matters in the array list.

File: src/TempleLineOfCredit.sol [hunk 12]
```diff
@@ -8,85 +8,115 @@ import {SafeERC20} from "openzeppelin-contracts/token/ERC20/utils/SafeERC20.sol"
...
 
-    /// @notice Total debt taken out
-    uint256 public debtBalance;
+    /// @notice mapping of debt token to its underlying information
+    mapping(address => DebtToken) public debtTokens;
 
-    /// @notice Fixed borrow interest rate in bpds
-    uint256 public immutable interestRateBps;
+    /// @notice list of all supported debtTokens
+    address[] public debtTokenList;
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:68]: Can you add a getter to return the debtTokenList?

```
function getDebtTokenList() external view returns (address[] memory) {
    return debtTokenList;
}
```

The dapp will need it.

File: src/TempleLineOfCredit.sol [hunk 13]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+        debtTokenList.push(token);
+        emit AddDebtToken(token);
     }
 
+
+    function removeDebtToken(address debtToken) external onlyOperators {
+        if (!debtTokens[debtToken].isAllowed) revert Unsupported(debtToken);
+        delete debtTokens[debtToken];
+        /// TODO: Remove from the debtTokenList
+        emit AddDebtToken(debtToken);
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: ...wrong event - needs to be `RemoveDebtToken`

Are you testing for this? You should also add tests to check the right events are emitted

File: src/TempleLineOfCredit.sol [hunk 14]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+    }
+
+    function setMinCollateralizationRatio(address debtToken, uint256 _minCollateralizationRatio) external onlyOperators {
+        if (!debtTokens[debtToken].isAllowed) revert Unsupported(debtToken);
+        debtTokens[debtToken].minCollateralizationRatio = _minCollateralizationRatio;
+        emit RemoveDebtToken(debtToken);
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Wrong event being emitted.

File: src/TempleLineOfCredit.sol [hunk 15]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
 
+
+    function removeDebtToken(address debtToken) external onlyOperators {
+        if (!debtTokens[debtToken].isAllowed) revert Unsupported(debtToken);
+        delete debtTokens[debtToken];
+        /// TODO: Remove from the debtTokenList
+        emit AddDebtToken(debtToken);
+    }
+
+    function setMinCollateralizationRatio(address debtToken, uint256 _minCollateralizationRatio) external onlyOperators {
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:167]: How do we set the other DebtToken parameters -- eg `interestRateBps`? That would need a removeDebtToken -> addDebtToken? Any other setters worth adding to set in place rather than two tx's?

File: src/TempleLineOfCredit.sol [hunk 16]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+        if (!debtTokens[debtToken].isAllowed) revert Unsupported(debtToken);
+        debtTokens[debtToken].minCollateralizationRatio = _minCollateralizationRatio;
+        emit RemoveDebtToken(debtToken);
+    }
+
+
     /**
      * @dev Get user principal amount
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Is this comment correct? It's not really the principal amount, more like the 'users latest actualised debt amount for a given debt token'. As it also increases over time too whenever the accrued interest is actualised.

Also the function name should be renamed to `getDebtPosition`, and made `external` (not used anywhere internally)

File: src/TempleLineOfCredit.sol [hunk 17]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
     /**
      * @dev Get user total debt incurred (principal + interest)
      * @return total Debt
      */
-    function getTotalDebtAmount(address account) public view returns (uint256) {
-        uint256 totalDebt = positions[account].debtAmount;
-        uint256 secondsElapsed = block.timestamp - positions[account].createdAt; 
-        totalDebt += (totalDebt * interestRateBps * secondsElapsed)  / 10000 / 365 days;
+    function getTotalDebtAmount(address debtToken, address account) public view returns (uint256) {
+        DebtToken memory debtTokenInfo = debtTokens[debtToken];
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: This is doing a copy unecessarily...would be better if it were storage, ie
```
DebtToken storage debtTokenInfo = debtTokens[debtToken];
```

File: src/TempleLineOfCredit.sol [hunk 18]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
 
+    function getTokenPrice(TokenPrice _price) public view returns (uint256 price, uint256 precision) {
+
+        if (_price == TokenPrice.STABLE) {
+            return (10000, 10000);
+        } else if (_price == TokenPrice.TPI) {
+            // Get Token Price from redemeer
+            uint256 tpiPrice = IOudRedeemer(oudRedeemer).treasuryPriceIndex();
+            return (tpiPrice, 10000);
+        }
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:215]: Should either have an extra `else` branch which reverts (unsupported token price type), or if you're ok assuming this will be updated if another option is added, just change above to an `else` to save an extra comparison.

File: src/TempleLineOfCredit.sol [hunk 19]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+        } else if (_price == TokenPrice.TPI) {
+            // Get Token Price from redemeer
+            uint256 tpiPrice = IOudRedeemer(oudRedeemer).treasuryPriceIndex();
+            return (tpiPrice, 10000);
+        } 
+    }
+
     /**
      * @dev Allows operator to depoist debt tokens
      * @param amount is the amount to deposit
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:222]: Typo in the comment `depoist`, also add the other params to the natspec (or don't include at all)

File: src/TempleLineOfCredit.sol [hunk 20]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
             account,
             address(this),
             amount
         );
-        emit DepositDebt(amount);
+        emit DepositReserve(debtToken, amount);
     }
 
         /**
      * @dev Allows operator to remove debt token
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:238]: Need to add the other @params or remove them entirely

File: src/TempleLineOfCredit.sol [hunk 21]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
-            amount
+            collateralAmount 
         );
-        emit PostCollateral(msg.sender, amount);
+        emit PostCollateral(msg.sender, collateralAmount);
     }
 
-    function borrow(uint256 borrowAmount) external {
-        if (borrowAmount == 0) revert InvalidAmount(borrowAmount);
+    function borrow(address[] memory tokens, uint256[] memory borrowAmounts) external {
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:275]: Needs a natspec comment on this function

File: src/TempleLineOfCredit.sol [hunk 22]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
-        position.collateralAmount -= collateralSeized;
-        position.debtAmount = 0;
-        position.createdAt = 0;
+        positions[debtor].collateralAmount -= collateralSeized;
+
+        // Wipe out all of users other token debts
+        address debtToken;
+        for (uint256 i; i < debtTokenList.length; ++i) {
+            debtToken = debtTokenList[i];
+            TokenPosition storage userPosition = positions[debtor].tokenPosition[debtToken];
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Forge shows a few ` Warning: This declaration shadows an existing declaration.`

Including this line. Can you clean them all up so `forge build` shows no warnings?

Declaration of the var can be moved above the for loop too.

File: src/TempleLineOfCredit.sol [hunk 23]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
         }
 
-        uint256 borrowCapacity = _maxBorrowCapacity(positions[msg.sender].collateralAmount) - debtAmount;
-        debtAmount += borrowAmount;
-
-        if (debtAmount > borrowCapacity) {
-            revert InsufficentCollateral(borrowCapacity, debtAmount);
+        address debtToken;
+        uint256 borrowAmount;
+        DebtToken memory debtTokenInfo;
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: This should be `storage` so it doesn't do an extra copy from memory.

File: src/TempleLineOfCredit.sol [hunk 24]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+            debtToken = tokens[i];
+            borrowAmount = borrowAmounts[i];
+
+            if (borrowAmount == 0) revert InvalidAmount(borrowAmount);
+
+            debtTokenInfo = debtTokens[debtToken];
+            if (!debtTokenInfo.isAllowed) revert Unsupported(debtToken); 
+
+            TokenPosition storage userPosition = positions[msg.sender].tokenPosition[debtToken];
+            uint256 debtAmount = userPosition.debtAmount;
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Optional - could be a little more concise and this is more gas efficient than first allocating and then updating
```
uint256 debtAmount = (userPosition.debtAmount == 0) ? 0 : getTotalDebtAmount(debtToken, msg.sender);
```

File: src/TempleLineOfCredit.sol [hunk 25]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+            
+            debtToken = tokens[i];
+            borrowAmount = borrowAmounts[i];
+
+            if (borrowAmount == 0) revert InvalidAmount(borrowAmount);
+
+            debtTokenInfo = debtTokens[debtToken];
+            if (!debtTokenInfo.isAllowed) revert Unsupported(debtToken); 
+
+            TokenPosition storage userPosition = positions[msg.sender].tokenPosition[debtToken];
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: The declaration of `userPosition`, `debtAmount`, `borrowCapacity` could be moved above the for loop

(even though `userPosition` is storage, I think it still works - but needs checking)

File: src/TempleLineOfCredit.sol [hunk 26]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
-            msg.sender,
-            borrowAmount 
-        );
-        emit Borrow(msg.sender, borrowAmount);
     }
 
-    function maxBorrowCapacity(address account) public view returns(uint256) {
-        return  _maxBorrowCapacity(positions[account].collateralAmount);
+    function maxBorrowCapacity(address debtToken, address account) public view returns(uint256) {
+        DebtToken memory debtTokenInfo = debtTokens[debtToken];
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Should be `storage` to avoid an unnecessary copy

File: src/TempleLineOfCredit.sol [hunk 27]
```diff
@@ -98,125 +128,184 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+            TokenPosition storage userPosition = positions[msg.sender].tokenPosition[debtToken];
+            uint256 debtAmount = userPosition.debtAmount;
+            if (debtAmount != 0) {
+                debtAmount = getTotalDebtAmount(debtToken, msg.sender);
+            }
+
+            uint256 borrowCapacity = _maxBorrowCapacity(positions[msg.sender].collateralAmount, debtTokenInfo.tokenPrice, debtTokenInfo.minCollateralizationRatio) - debtAmount;
+            debtAmount += borrowAmount;
+
+            if (debtAmount > borrowCapacity) {
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: I'm not sure this check is correct given how you've written it? I could definitely be wrong, but let's say:

* max borrow capacity = 1_000 DAI
* existing debt amount = 900 DAI
* new requested borrow amount = 50 DAI

Then I would expect it to be ok as the user could still borrow up to 100 more DAI

However with your logic:
```
borrowCapacity = 1000 - 900 = 100
debtAmount = 900 + 50 = 950
950 > 100, so it would revert.
```

Have you got tests for this part? Can you stress any edge cases?

File: src/TempleLineOfCredit.sol [hunk 28]
```diff
@@ -230,10 +319,17 @@ contract TempleLineOfCredit is Ownable, Operators {
             revert ExceededCollateralAmonut(msg.sender, collateralAmount, withdrawalAmount);
         }
 
-        uint256 borrowCapacity = _maxBorrowCapacity(collateralAmount - withdrawalAmount);
-        
-        if (positions[msg.sender].debtAmount > borrowCapacity ) {
-            revert WillUnderCollaterlize(msg.sender, withdrawalAmount);
+        for (uint256 i; i < debtTokenList.length; ++i) {
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: This is an extra sload on each iteration because it's storage. Probably better to do:

```
uint256 debtTokensLength = debtTokenList.length;
for (uint256 i; i < debtTokensLength; ++i) {
```

File: src/TempleLineOfCredit.sol [hunk 29]
```diff
@@ -230,10 +319,17 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+
+            address debtToken = debtTokenList[i];
+            DebtToken memory debtTokenInfo = debtTokens[debtToken];
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: 1. Should be storage
2. Can move these param declarations above the for loop

File: src/TempleLineOfCredit.sol [hunk 30]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
-        if (repayAmount == 0) revert InvalidAmount(repayAmount);
-        positions[msg.sender].debtAmount = getTotalDebtAmount(msg.sender);
-
-        if (repayAmount >  positions[msg.sender].debtAmount) {
-            revert ExceededBorrowedAmount(msg.sender, positions[msg.sender].debtAmount, repayAmount);
+    function repay(address[] memory tokens, uint256[] memory repayAmounts) external {
+
+        address debtToken;
+        uint256 repayAmount;
+        DebtToken memory debtTokenInfo;
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Should be `storage` to avoid a copy

File: src/TempleLineOfCredit.sol [hunk 31]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+            debtToken = tokens[i];
+            repayAmount = repayAmounts[i];
+            
+            if (repayAmount == 0) revert InvalidAmount(repayAmount);
+
+            debtTokenInfo = debtTokens[debtToken];
+            if (!debtTokenInfo.isAllowed) revert Unsupported(debtToken); 
+
+
+            TokenPosition storage userPosition = positions[msg.sender].tokenPosition[debtToken];
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: More params to move above the for loop

File: src/TempleLineOfCredit.sol [hunk 32]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+            uint256 debtAmount = userPosition.debtAmount;
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]:
```
uint256 debtAmount = (userPosition.debtAmount == 0) ? 0 : getTotalDebtAmount(debtToken, msg.sender);
```

File: src/TempleLineOfCredit.sol [hunk 33]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
 
    /**
      * @dev Allows borrower to repay borrowed amount
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:376]: Because the debt amount is increasing every single second, it's very very hard for users to pay off the entire amount with just this function - as the block.timestamp keeps increasing...

I think there may also need to be a `repayAll(address[] memory tokens)` in order to faciliate this?

File: src/TempleLineOfCredit.sol [hunk 34]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+            userPosition.lastUpdatedAt = block.timestamp;
+            
+            if (debtTokenInfo.tokenType == TokenType.TRANSFER){
+                IERC20(debtToken).safeTransferFrom(
+                    msg.sender,
+                    address(this),
+                    repayAmount 
+                );
+            } else {
+                IERC20MintBurn(debtToken).burn(
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:415]: This assumes this contract has burn rights over someone else's tokens. I think that's probably the case for OUD, but it isn't for say $TEMPLE (a user can burn their own temple, but our MSIG can't burn their temple)

File: src/TempleLineOfCredit.sol [hunk 35]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
      * @dev Allows operator to liquidate debtors position
      * @param debtor the account to liquidate
      */
-    function liquidate(address debtor) external onlyOperators {
-        Position storage position = positions[debtor];    
-        uint256 totalDebtOwed = getTotalDebtAmount(debtor);
+    function liquidate(address debtor, address debtToken) external onlyOperators {
 
-        if (_getCurrentCollaterilizationRatio(position.collateralAmount, position.debtAmount, totalDebtOwed) >= minCollateralizationRatio) {
+        DebtToken memory debtTokenInfo = debtTokens[debtToken];
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Should be storage to save a copy

The line below should use this `debtTokenInfo` too...

File: src/TempleLineOfCredit.sol [hunk 36]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+        if (!debtTokens[debtToken].isAllowed) revert Unsupported(debtToken);
+        
+        TokenPosition storage userPosition = positions[debtor].tokenPosition[debtToken];
+        uint256 totalDebtOwed = userPosition.debtAmount;
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]:
```
uint256 totalDebtOwed = userPosition.debtAmount == 0 ? 0 : getTotalDebtAmount(debtToken, debtor);
```

And can then remove the `if (totalDebtOwed != 0)` check below

File: src/TempleLineOfCredit.sol [hunk 37]
```diff
@@ -299,15 +441,18 @@ contract TempleLineOfCredit is Ownable, Operators {
         emit Liquidated(debtor, totalDebtOwed, collateralSeized);
     }
 
-    function getCurrentCollaterilizationRatio(address account) public view returns(uint256) {
-        _getCurrentCollaterilizationRatio(positions[account].collateralAmount, positions[account].debtAmount, getTotalDebtAmount(account));
+    function getCurrentCollaterilizationRatio(address debtToken, address account) public view returns(uint256) {
+        DebtToken memory debtTokenInfo = debtTokens[debtToken];
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: This should be storage to avoid a copy. Probably not required though - can just do inline in the line below, as it's only used once.

Also the function should be external, not public

File: src/TempleLineOfCredit.sol [hunk 38]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+        if (totalDebtOwed != 0) {
+                totalDebtOwed = getTotalDebtAmount(debtToken, debtor);
+        }
+
+        if (_getCurrentCollaterilizationRatio(collateralAmount, userPosition.debtAmount, debtTokenInfo.tokenPrice,  totalDebtOwed) >= debtTokenInfo.minCollateralizationRatio) {
             revert OverCollaterilized(debtor);
         }
 
-        uint256 collateralSeized = (totalDebtOwed * debtPrice) / collateralPrice;
+        (uint256 debtTokenPrice, uint256 debtPrecision) = getTokenPrice(debtTokenInfo.tokenPrice);
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:438]: These prices are also calculated within `_getCurrentCollaterilizationRatio()`. Given this will be a DAO-paid function and may run on many accounts, it's worth optimising (given this does external calls).

You could do this above, and then pass the prices/precisions into `_getCurrentCollaterilizationRatio`

File: src/TempleLineOfCredit.sol [hunk 39]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+        (uint256 collateralTokenPrice, uint256 collateralPrecision) = getTokenPrice(collateralPrice);
+
 
-        if (collateralSeized > position.collateralAmount) {
-            collateralSeized = position.collateralAmount;
+        uint256 collateralSeized = (totalDebtOwed * debtTokenPrice * collateralPrecision) / collateralTokenPrice / debtPrecision;
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol:445]: This doesn't have the extra `* 10_000` like the others. Should it?

File: src/TempleLineOfCredit.sol [hunk 40]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+
+        if (collateralSeized > collateralAmount) {
+            collateralSeized = collateralAmount;
         }
 
-        position.collateralAmount -= collateralSeized;
-        position.debtAmount = 0;
-        position.createdAt = 0;
+        positions[debtor].collateralAmount -= collateralSeized;
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: An unnecessary sload here since it's storage. Better as
```
positions[debtor].collateralAmount = collateralAmount - collateralSeized
```
since you already have it in memory

File: src/TempleLineOfCredit.sol [hunk 41]
```diff
@@ -247,49 +343,95 @@ contract TempleLineOfCredit is Ownable, Operators {
...
+
+        // Wipe out all of users other token debts
+        address debtToken;
+        for (uint256 i; i < debtTokenList.length; ++i) {
```
Comment (frontier159) [inline src/TempleLineOfCredit.sol]: Save sload's each iteration:
```
uint256 debtTokenLength = debtTokenList.length;
for (uint256 i; i < debtTokenLength; ++i) {
```

**Review by frontier159 — CHANGES_REQUESTED** [review]

File: src/TempleLineOfCredit.sol (hunk 7 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol]: 👍

File: src/TempleLineOfCredit.sol (hunk 10 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol:21]: this will be done in a separate PR

File: src/TempleLineOfCredit.sol (hunk 39 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol:445]: no the math checks out here

File: src/TempleLineOfCredit.sol (hunk 38 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol:438]: 👍

File: src/TempleLineOfCredit.sol (hunk 12 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol:68]: 👍

File: src/TempleLineOfCredit.sol (hunk 18 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol:215]: 👍

File: src/TempleLineOfCredit.sol (hunk 26 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol]: 👍

File: src/TempleLineOfCredit.sol (hunk 34 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol:415]: yeah so this is meant for OUD tokens. TLC should have min/burn permission

File: src/TempleLineOfCredit.sol (hunk 33 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol:376]: hmm, 🤔  thats a good point. Let me add it in a separate PR

File: src/TempleLineOfCredit.sol (hunk 24 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol]: 👍

File: src/TempleLineOfCredit.sol (hunk 25 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol]: 👍

File: src/TempleLineOfCredit.sol (hunk 27 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol]: oops brain fart should be ` if (debtAmount > borrowCapacity) {` 👍

Comment (shero0x1337) [reply src/TempleLineOfCredit.sol]: fixed and added assertion test

File: src/TempleLineOfCredit.sol (hunk 9 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol:88]: yeah ack. Let me think about the events i want to index a little bit more. Will add in a separate pr.

File: src/TempleLineOfCredit.sol (hunk 15 above)
Comment (shero0x1337) [reply src/TempleLineOfCredit.sol:167]: @frontier159 wanted to think about the ramification allowing the changing of this variables a little bit. Ideally this should not be getting changed on the fly. Like if you change the `interestRateBps` it will also affect the interest rate that were accrued from a while ago since we accure debt on demand. Which is unfair to the user

**Comment by shero0x1337** [comment]
Merging for now to consolidate branches
